sample-app/src/components/InputDropdown.tsx

```diff
@@ -52,74 +47,120 @@ export default function InputDropdown({
   placeholder,
   screenReaderInstructions,
   screenReaderInstructionsId,
-  options,
-  optionIdPrefix,
+  screenReaderText,
+  children,
   onSubmit = () => {},
-  updateInputValue,
-  updateDropdown,
+  onInputChange,
+  onInputFocus,
   renderButtons = () => null,
   cssClasses
-}: Props): JSX.Element | null {
+}: React.PropsWithChildren<Props>): JSX.Element | null {
   const [{
-    focusedOptionIndex,
+    focusedSectionIndex,
     shouldDisplayDropdown,
   }, dispatch] = useReducer(reducer, {
-    focusedOptionIndex: undefined,
+    focusedSectionIndex: undefined,
     shouldDisplayDropdown: false,
   });
-  const focusOptionId = focusedOptionIndex === undefined
-    ? undefined
-    : `${optionIdPrefix}-${focusedOptionIndex}`;
 
+  const [focusedOptionId, setFocusedOptionId] = useState<string | undefined>(undefined);
   const [latestUserInput, setLatestUserInput] = useState(inputValue);
   const [screenReaderKey, setScreenReaderKey] = useState(0);
 
   const inputRef = useRef<HTMLInputElement>(document.createElement('input'));
+  const dropdownRef = useRef<HTMLDivElement>(document.createElement('div'));
 
   if (!shouldDisplayDropdown && screenReaderKey) {
     setScreenReaderKey(0);
   }
 
+  const childrenArray = React.Children.toArray(children);
+  const childrenWithProps = childrenArray.map((child, index) => {
+    if (!React.isValidElement(child)) {
+      return child;
+    }
+    
+    const modifiedOnClickOption : typeof child.props.onClickOption = (option: typeof child.props.options[0]) => {
+      child.props.onClickOption(option);
+      dispatch({ type: 'HideSections' });
+    }
+
+    const modifiedOnFocusChange = (value: string, focusedOptionId: string) => {
+      child.props.onFocusChange(value);
+      setFocusedOptionId(focusedOptionId);
+    };
+
+    if (focusedSectionIndex === undefined) {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'reset', onClickOption: modifiedOnClickOption });
+    } else if (index === focusedSectionIndex) {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'active', onFocusChange: modifiedOnFocusChange, onClickOption: modifiedOnClickOption });
+    } else {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'inactive', onClickOption: modifiedOnClickOption });
+    }
+  });
+
+  const numSections = childrenWithProps.length;
+
+  function onLeaveSectionFocus(focusNext: boolean) {
+    if (focusedSectionIndex === undefined && focusNext) {
+      dispatch({ type: 'FocusSection', newIndex: 0 });
+    } else if (focusedSectionIndex !== undefined && focusedSectionIndex < numSections - 1) {
```

Comment:
I think focusedSectionIndex !== undefined can be removed

Reply:
I restructured it, so now I think the `focusedSectionIndex !== undefined` should stay